### PR TITLE
Organize CMake and add Coverage build type

### PR DESCRIPTION
**Pull request by Warchant**
- Organize cmake in a better structure - analysis, config types. Shared model project can re-use that config.
- Add new `CMAKE_BUILD_TYPE=Coverage` -- it installs all necessary flags for coverage and should be used in CI.
- Remove unused fuzzing
- Remove COVERAGE option -- instead, comparison of ${CMAKE_BUILD_TYPE} with `Coverage` can be used.
- Organize flags
- Add ccache support for xcode, ninja, and linker

Same as https://github.com/hyperledger/iroha/pull/983 but without unnecessary changes. All changes are necessary, tested and related to CMake organization.

### CODE REVIEW

**Comment by Warchant** [comment]
@lebdron why you don't like addition of new build type? It is intuitive that build type is kind of configuration in cmake, since we have at least multiple configurations (Debug,Release + Coverage), it is wise to add new config.

Debug + COVERAGE = redundant -g -O0 flags. We need -O1 to collect better coverage, and we do not need -g, it is redundant for coverage (it just consumes memory and increases build time).
Release + COVERAGE = -NDEBUG is a redundant flag (it disables asserts, but it is preferable to enable them, since asserts may contain execution paths which influence program state), plus `build` step in CI should be compiled in Release in this case, as well as tests. 

Coverage type is mostly for CI, you will not use it manually. You will typically build iroha in Debug or Release.

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,40 +1,31 @@
...
+#########################
+set(CMAKE_CONFIGURATION_TYPES "Debug;Release;Coverage" CACHE STRING "" FORCE)
 
-SET(CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/cmake/Modules)
+
+######################
+## project settings ##
+######################
+
+PROJECT(
```
Comment (l4l) [inline CMakeLists.txt]: Use just `project(iroha LANGUAGES C CXX)` (note еру case)
Also there is some other problems with the case below (like JOIN), check that plz

File: cmake/analysis/gcovr.cmake [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  message(STATUS "Target gcovr enabled")
+
+  if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
+    set(GCOV_BACKEND "llvm-cov gcov")
+  else()
+    set(GCOV_BACKEND "gcov")
+  endif()
+
+  add_custom_target(gcovr
+      COMMAND ${GCOVR_BIN} -s -x -r '${CMAKE_SOURCE_DIR}'
```
Comment (l4l) [inline cmake/analysis/gcovr.cmake]: Use long flags, like: `--print-summary --xml --root=${CMAKE_SOURCE_DIR} --exclude=...`
much easier to understand what is happening

File: CMakeLists.txt (hunk 1 above)
Comment (Warchant) [reply CMakeLists.txt]: okay. lowercase-all-the-things

File: cmake/analysis/gcovr.cmake (hunk 2 above)
Comment (Warchant) [reply cmake/analysis/gcovr.cmake]: Nice suggestion, but it was previously like this.

**Comment by muratovv** [comment]
The PR contains a lot of independent changes. Need to separate it into logically independent PRs for continue. The task about it is here https://soramitsu.atlassian.net/browse/IR-957.

**Pull request closed without merging**